### PR TITLE
Two fixes for /metadata 

### PR DESCRIPTION
**Pull request by mossprescott**
These turned up in looking at SD-1331, but may not entirely account for what's going on there.

### CODE REVIEW

File: core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala [hunk 1]
```diff
@@ -161,9 +161,15 @@ private final class QueryFileInterpreter[C](
       }).liftM[QRT].liftM[WorkflowExecErrT]
 
     case FileExists(file) =>
-      collFromPathM(file)
-        .flatMap(MongoDbIO.collectionExists(_).liftM[FileSystemErrT])
-        .run.liftM[QRT].liftM[WorkflowExecErrT]
+      // NB: there can never be a collection at a path directly under the root
+      // (only a database), but it is not an error to ask the question.
```
Comment (jedesah) [inline core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala]: :+1: For a clear and concise comment!

File: core/src/test/scala/quasar/fs/mount/view.scala [hunk 2]
```diff
@@ -347,12 +347,16 @@ class ViewFSSpec extends Specification with ScalaCheck with TreeMatchers {
...
+        viewInterp(Views.empty, paths, program) must_==((ops, expected))
       }
       val noFile = {
-        val expected = (ops, false.right)
-        viewInterp(Views.empty, Map(), program) must_== expected
+        val expected =
+          if (fileParent(file) ≟ rootDir)
+            false.right
+          else
+            FileSystemError.pathError(PathError2.pathNotFound(fileParent(file))).left
```
Comment (jedesah) [inline core/src/test/scala/quasar/fs/mount/view.scala]: I'm getting a little tired and will probably take another fresh look at this tmr morning. But it strikes me as odd that we expect a `PathNotFound` as a return value to `FileExists`, especially since in the `MongoDbFileSystemSpec` below we seem to be expecting `false.right` even though the parent directory does not exists which leads me to believe that the desired behavior for `FileExists` is to always return `true` or `false`, but never `NotFound`.

Comment (mossprescott) [inline core/src/test/scala/quasar/fs/mount/view.scala]: I think it's a good question.

For the MongoDB FS, I think clearly the right answer is always `false` and never `pathNotFound`, because in principle any path at all _could_ contain a file, so it's not like something about the path is making it impossible to answer the question. By the way, it may not be obvious, but the change to MongoDb is to stop it producing `invalidPath`, so I still that's legit regardless.

On the other hand, for `hierarchical`, it may make sense to produce `pathNotFound` in the case that the path is outside any mounted filesystem, which is what's happening here. That would make it clear that in the current configuration, no file could _ever_ appear at that path.

On the third hand, I'm pretty sure we actually would want the API to produce a 404 in that case, and the kicker is that it does, because `false` actually gets mapped to `pathNotFound` at that level.

So at the moment it's academic. However, given that the type is `FileSystemError \/ Boolean` (and we probably need to keep that, in case of other errors), I don't think there's a good way to express that this method doesn't produce `pathNotFound`, unless of course we want to go with something like `assertFileExists(file: AFile): FileSystemError \/ Unit` (or `Option[FileSystemError]`, as you're probably thinking ;).

Also, one way or the other, we still need the view interpreter to handle these error cases, because the idea is it's possible to mount views literally anywhere, regardless of what problems those paths might cause for the other interpreters. So I think this bug fix is mostly orthogonal to the issue of whether not this particular situation is handled this way.

File: core/src/main/scala/quasar/fs/mount/view.scala [hunk 3]
```diff
@@ -189,7 +189,14 @@ object view {
           })
 
         case FileExists(file) =>
-          query.fileExists(file).map(_ || views.contains(file)).run
+          query.fileExists(file).run.map(_ match {
+            case \/-(x) =>
+              (x || views.contains(file)).right
+            case -\/(err @ PathError(PathNotFound(_))) =>
```
Comment (mossprescott) [inline core/src/main/scala/quasar/fs/mount/view.scala]: After writing my response to @jedesah's comment, I'm actually thinking this should catch _all_ errors. Regardless of what problems the underlying interpreter has with the path, if there's a view mounted, there's a file there as far as this filesystem is concerned (because all references to that path will get re-written by this interpreter.)

Therefore I'll probably change this to a simple `fold` that handles all errors the same way. Make sense?

File: core/src/test/scala/quasar/fs/mount/view.scala (hunk 2 above)
Comment (wemrysi) [inline core/src/test/scala/quasar/fs/mount/view.scala]: I'm still not convinced the return type needs to be `FileSystemError \/ Boolean`, I think we could just return a `Boolean`. If a particular implementation wants to interpret into an effect with an implementation-specific error type, it can, but from the view of the filesystem abstraction I don't see how this could possibly fail.

Comment (jedesah) [inline core/src/test/scala/quasar/fs/mount/view.scala]: I can get behind that. This feels like deja vu. Did we already have a discussion on this topic before? Does anyone remember where that is?

Comment (wemrysi) [inline core/src/test/scala/quasar/fs/mount/view.scala]: We discussed it previously in #1044

Comment (jedesah) [inline core/src/test/scala/quasar/fs/mount/view.scala]: @wemrysi Thx for the link. That helps me refresh my mind on the issue. I think having it return just `Boolean` is probably the right answer. I do mention in that thread that we should make sure the behavior is coherent, i.e. that we should make sure all mongo specific errors are inserted in the mongodb interpreter as opposed to the term result.

File: core/src/main/scala/quasar/fs/mount/view.scala (hunk 3 above)
Comment (jedesah) [inline core/src/main/scala/quasar/fs/mount/view.scala]: That makes sense to me.

File: core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala (hunk 1 above)
Comment (jedesah) [inline core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala]: I wonder though, if we add `FileSystemErrT` to the mongodb interpreter monad stack if it makes sense to put this error in there instead of returning `false` conveying in some way the fact that it is IMPOSSIBLE for there to ever be a file here.

File: core/src/test/scala/quasar/fs/TraceFS.scala [hunk 4]
```diff
@@ -48,9 +56,7 @@ object TraceFS {
           case Explain(lp)          => (Vector.empty, \/-(ExecutionPlan(FsType, lp.toString)))
           case ListContents(dir)    => ls(dir)
           case FileExists(file)     =>
-            ls(fileParent(file)).fold(
-              κ(false),
-              _.contains(fileName(file).right)).right
+            ls(fileParent(file)).map(_.contains(fileName(file).right))
```
Comment (jedesah) [inline core/src/test/scala/quasar/fs/TraceFS.scala]: What is the rationale behind this change?

Comment (mossprescott) [inline core/src/test/scala/quasar/fs/TraceFS.scala]: It makes it possible to test the behavior of the view interpreter when the underlying interpreter fails. As it was, it was always hiding those failures by converting them to `false.right`.

File: it/src/test/scala/quasar/physical/mongodb/fs/MongoDbFileSystemSpec.scala [hunk 5]
```diff
@@ -280,6 +280,24 @@ class MongoDbFileSystemSpec
         }
       }
 
+      "File exists" >> {
+        "for missing file at root (i.e. a database path) should succeed" >> {
+          val tfile = rootDir </> file("foo")
+
+          val p = query.fileExists(tfile)
+
+          runT(run)(p).runEither must beRight(false)
```
Comment (jedesah) [inline it/src/test/scala/quasar/physical/mongodb/fs/MongoDbFileSystemSpec.scala]: So, as per my comment above, we should consider whether this should yield a left since it is a MongoDB particularity. Maybe that's not a good idea, just wanted to put a comment to remember to change this if we go that route.

File: core/src/test/scala/quasar/fs/TraceFS.scala [hunk 6]
```diff
@@ -23,7 +23,15 @@ import quasar.fp._
 import pathy.{Path => PPath}, PPath._
 import scalaz._, Scalaz._
 
-/** Interpreter that just records a log of the actions that are performed. */
+/** Interpreter that just records a log of the actions that are performed. Only a
```
Comment (jedesah) [inline core/src/test/scala/quasar/fs/TraceFS.scala]: Total side comment, but I have wondered in the past what `TraceFS` does more than `Free` since `Free` is also something that "records" what needs to be done. Maybe `Free` records what needs to be done and `TraceFS` records what was actually done.

Just wondering if it would be possible to compare different instances of `Free[S,A]` for equality instead of interpreting them and inspecting the traces.

File: core/src/main/scala/quasar/fs/mount/view.scala (hunk 3 above)
Comment (jedesah) [inline core/src/main/scala/quasar/fs/mount/view.scala]: Interestingly though, if we changed the result type of `FileExists` to `Boolean` instead of `FileSystemError \/ Boolean`, this would make it harder for the `view` interpreter to "catch" problems with an underlying interpreter, but return `true` as long as there is a view that's mounted there. Or would it?

File: core/src/test/scala/quasar/fs/TraceFS.scala (hunk 4 above)
Comment (jedesah) [inline core/src/test/scala/quasar/fs/TraceFS.scala]: Ah, that makes sense.

File: core/src/main/scala/quasar/fs/mount/view.scala (hunk 3 above)
Comment (mossprescott) [inline core/src/main/scala/quasar/fs/mount/view.scala]: I was just writing a reply on that very question.

I think for FileExists there's an easy solution to that, which is that the view interpreter can simply check for a view _first_, and not call the underlying interpreter if it can return `true` from it's own information.

Comment (jedesah) [inline core/src/main/scala/quasar/fs/mount/view.scala]: That makes a lot of sense. Too much sense actually, I am jealous I did not think of that myself!

Comment (mossprescott) [inline core/src/main/scala/quasar/fs/mount/view.scala]: Does a similar argument apply to ListContents?

1) If an error occurs the prevents the mongodb interpreter from determining the contents of a directory (say, a connection failure), then it can produce an error in some Failure effect. The view interpreter wouldn't be able to intercept that, but that's OK because it would have no way of producing a correct result anyway.

2) if a path is constructed in such a way that it could never have children _in MongoDB_, then the interpreter would have to produce a value so as to allow the view interpreter to overlay any views (which can exist at _any_ path). But what value?

Right now, the answer is `pathNotFound.left: FileStsemError \/ Set[PathName]`, but another option would be `none: Option[Set[PathName]]` and yet another would be simply `Set[Path]()`. I'm not sure I like any of those all that much.

File: core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala (hunk 1 above)
Comment (wemrysi) [inline core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala]: If we're interested in conveying this sort of information, maybe we should encode it into the return type of `FileExists` (and probably rename `FileExists` to something else to communicate the additional behavior). I.e. return a type that has cases like "exists", "doesn't exist", "couldn't exist", etc.

I'm not sure if we need this, but we could then implement `fileExists` as a derviative of this more precise operation that just throws away information, collapsing "doesn't exist" and "couldn't exist" into `false`.

It doesn't seem like asking if a path that could never exist exists should result in an error though, when `false` is an accurate answer.

File: core/src/main/scala/quasar/fs/mount/view.scala (hunk 3 above)
Comment (wemrysi) [inline core/src/main/scala/quasar/fs/mount/view.scala]: What is undesirable about the view filesystem handling the `pathNotFound` returned by the underlying filesystem?

File: core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala (hunk 1 above)
Comment (jedesah) [inline core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala]: I guess that could be a solution. The disadvantage of it being that it does not allow different interpreters to convey whether THEY have any special cases or not and what the nature of the special cases unlike in the original solution you proposed of having it be part of the interpreter monad stack.

File: core/src/main/scala/quasar/fs/mount/view.scala (hunk 3 above)
Comment (mossprescott) [inline core/src/main/scala/quasar/fs/mount/view.scala]: I'm just asking whether `FileSystemError \/ Set[PathName]` makes sense as a type, but I'm willing to drop it ;)

File: core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala (hunk 1 above)
Comment (wemrysi) [inline core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala]: I guess what I'm trying to get at is that it seems surprising, as a client of the API, to receive an error when asking if a file exists because a particular implementation happens to know it could never exist, when `false` would answer the question asked adequately.

File: core/src/main/scala/quasar/fs/mount/view.scala (hunk 3 above)
Comment (wemrysi) [inline core/src/main/scala/quasar/fs/mount/view.scala]: It seems like `FileSystemError \/ Set[PathName]` is an improvement over `Set[PathName]` as the latter conflates an empty directory with one that doesn't exist, whereas the former allows us to communicate that the operation is only defined for existing directories.

Comment (mossprescott) [inline core/src/main/scala/quasar/fs/mount/view.scala]: Yeah, kinda. Anyway, like I said, I'm over it.

File: core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala (hunk 1 above)
Comment (jdegoes) [inline core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala]: > I guess what I'm trying to get at is that it seems surprising, as a client of the API, to receive an error when asking if a file exists because a particular implementation happens to know it could never exist, when false would answer the question asked adequately.

:+1: An existence API should never error merely because a file cannot exist. 

Think of it from the perspective of the SlamData app (a prototypical Quasar app): the app wants to check if a file exists and will branch off the true/false return value. But if suddenly, some "falses" will be encoded by an error, then the app has to know that fact, and then internally capture and decode the error and treat it as equivalent to the "false" case.

File: core/src/test/scala/quasar/fs/TraceFS.scala (hunk 6 above)
Comment (mossprescott) [inline core/src/test/scala/quasar/fs/TraceFS.scala]: That is a thought-provoking question, and not one I can readily answer. I think the practical fact is that you cannot compare the two things we have in these tests, which in general are a `Free[FileSystem, A]` and a `Free[ViewFileSystem, A]`, because types (even if you could deal with the extra effects in the view program). But I'm sure there is a more subtle answer, along the lines of to be able to compare them, you would have to impose constraints X, Y, and Y…

**Comment by jdegoes** [comment]
> Total side comment, but I have wondered in the past what TraceFS does more than Free since Free is also something that "records" what needs to be done. Maybe Free records what needs to be done and TraceFS records what was actually done.

`Free` describes a program in `F`. The structure of the program is dynamic, and may change as it's being executed. In order to explore its structure, you need to execute it -- which corresponds to feeding it the values that it's reading from its environment. If you feed different values, you may get different programs.

In general, `Free[F, A]` can only be compared with `Free[F, A]`, of course, you can compare with `Free[G,A]` if you have `G ~> F` (i.e. `F` is "bigger" than `G`).

However, to perform the comparison, you must actually execute the program by feeding it values. Since the values you feed it determine the program's continuation, any comparison you do would be valid *strictly* for the specific values you fed it. 

So in the best case, you can compare `Free[F, A]` with `Free[F, A]` and determine that they yield the same `A` for the same input values. But to say they represent the same program, you have to prove this **for all possible input values**. Since that is not possible, you cannot prove that `Free[F, A]` is equal to another `Free[F, A]`, at least for some arbitrary `F`.

Indeed, you should expect this, because in general, two programs cannot be compared for equality. Even more basic, two functions cannot (in general) be compared for equality, because to do so you have to perform a symbolic proof, which corresponds to traversing an infinite graph by an unbounded distance.

However, that said, for specific `F` you may be able to prove equality. In particular, an `F` which does not read any values. This corresponds to a program whose structure does not depend on the result of computation; in some sense you can think of it as a "write only" program. In the Haskell formulation of `Free`, you can actually just pattern match to perform the analysis.

Meanwhile, however, you can execute a program, log its series of operations, and compare that log with another log. This is equivalent to a unit test which compares that two programs "did the same things in the same order". Of course, it is *not* a prove that two programs are the same, only a proof that for some fixed inputs, two programs produced the same set of operations in the same order.

Hence, `TraceFS` records what a program does, as it's doing it (according to some other interpreter responsible for producing the values the program needs), while `Free` merely describes the program generically in terms of some `F`.

**Comment by jedesah** [comment]
@jdegoes That makes sense. Thx for the explanation!

**Comment by jdegoes** [comment]
No problem!

**Comment by mossprescott** [comment]
Switching to just `Boolean` mostly simplified things a bit, but I did add a variant, `fileExistsM` that lifts the result into `FileSystemErrT`, because it was being used that way in a lot of places. This seems gross to me, but it's certainly not out of line with similar functions we have all over the place. 

Other than questions of code smell, it essentially didn't affect the APIs behavior, and all in all, I do think it's an improvement.

File: core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala [hunk 7]
```diff
@@ -161,9 +161,9 @@ private final class QueryFileInterpreter[C](
       }).liftM[QRT].liftM[WorkflowExecErrT]
 
     case FileExists(file) =>
-      collFromPathM(file)
-        .flatMap(MongoDbIO.collectionExists(_).liftM[FileSystemErrT])
-        .run.liftM[QRT].liftM[WorkflowExecErrT]
+      Collection.fromPathy(file).fold(
```
Comment (jedesah) [inline core/src/main/scala/quasar/physical/mongodb/fs/queryfile.scala]: I am still somewhat concerned with this behavior, i.e. simply returning `false` when there is something wrong with the path being provided. Either way, I think it's probably good that it's not part of the core algebra but it might make sense for the mongodb interpreter to report this as a failure in the monad stack rather than simply returning `false`. But we also already discussed this, so I can get on board with just returning `false`.

File: it/src/test/scala/quasar/physical/mongodb/fs/MongoDbFileSystemSpec.scala [hunk 8]
```diff
@@ -280,6 +280,24 @@ class MongoDbFileSystemSpec
         }
       }
 
+      "File exists" >> {
+        "for missing file at root (i.e. a database path) should succeed" >> {
+          val tfile = rootDir </> file("foo")
+
+          val p = query.fileExists(tfile)
+
+          run(p).run must_== false
```
Comment (jedesah) [inline it/src/test/scala/quasar/physical/mongodb/fs/MongoDbFileSystemSpec.scala]: So, just to confirm one last time, if someone uses the `metadata` api to query such a file in a mongodb filesystem, we are content returning `false` instead of `"It is not possible to have a file at the top level of a MongoDB filesystem"`?

File: core/src/test/scala/quasar/fs/TraceFS.scala [hunk 9]
```diff
@@ -50,7 +57,7 @@ object TraceFS {
           case FileExists(file)     =>
             ls(fileParent(file)).fold(
               κ(false),
-              _.contains(fileName(file).right)).right
+              _.contains(fileName(file).right))
```
Comment (jedesah) [inline core/src/test/scala/quasar/fs/TraceFS.scala]: I wonder if we should just change the implementation of `ls` here. Kind of weird to be constructing a `pathNotFound` just to ignore it and I doubt it's being used elsewhere, but I could be wrong.

Comment (jedesah) [inline core/src/test/scala/quasar/fs/TraceFS.scala]: Nevermind! We use it on line `56`. :facepunch:

File: core/src/test/scala/quasar/fs/mount/view.scala [hunk 10]
```diff
@@ -341,30 +341,24 @@ class ViewFSSpec extends Specification with ScalaCheck with TreeMatchers {
...
     "return true if there is a view at that path" ! prop { (file: AFile, lp: Fix[LogicalPlan]) =>
-      val program = query.fileExists(file).run
+      val program = query.fileExists(file)
 
-      val ops = traceInterp(program, Map())._1
-
-      val expected = (ops, true.right)
-
-      viewInterp(Views(Map(file -> lp)), Map(), program) must_== expected
+      viewInterp(Views(Map(file -> lp)), Map(), program) must_==((Vector(), true))
```
Comment (jedesah) [inline core/src/test/scala/quasar/fs/mount/view.scala]: I don't understand why we don't get the trace of `FileExists`?

File: it/src/test/scala/quasar/physical/mongodb/fs/MongoDbFileSystemSpec.scala (hunk 8 above)
Comment (mossprescott) [inline it/src/test/scala/quasar/physical/mongodb/fs/MongoDbFileSystemSpec.scala]: That's right. And from the outsider's point of view, it's possible for a file to appear there, if it's a view.

File: core/src/test/scala/quasar/fs/mount/view.scala (hunk 10 above)
Comment (mossprescott) [inline core/src/test/scala/quasar/fs/mount/view.scala]: When there's a view present, there's no longer any need to check the underlying FS.

File: it/src/test/scala/quasar/physical/mongodb/fs/MongoDbFileSystemSpec.scala (hunk 8 above)
Comment (jedesah) [inline it/src/test/scala/quasar/physical/mongodb/fs/MongoDbFileSystemSpec.scala]: :+1:

File: core/src/test/scala/quasar/fs/mount/view.scala (hunk 10 above)
Comment (jedesah) [inline core/src/test/scala/quasar/fs/mount/view.scala]: Ah, makes sense.

**Comment by jedesah** [comment]
LGTM!

**Comment by mossprescott** [comment]
Merged as 4.4.5 via the script.
